### PR TITLE
Finished assignment 3. Waiting for approval

### PR DESCRIPTION
**Pull request by andrefnb**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -34,4 +34,6 @@ jobs:
         pytest life_expectancy --cov
     - name: Lint with pylint
       run: |
-        pylint life_expectancy.cleaning
+        pylint life_expectancy.main
+        pylint life_expectancy.data_loading
+        pylint life_expectancy.data_cleaning
```
Comment (MrCordeiro) [inline .github/workflows/test.yml]:
```suggestion
        pylint life_expectancy
```

Incluira tamb]em os testes.

File: .vscode/settings.json
Comment (MrCordeiro) [inline .vscode/settings.json:1]: Não boa idea colocar esse fecheiro no commit.

File: life_expectancy/main.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+"""
+Main script for life_expectancy module
+"""
+
+from pathlib import Path
+from life_expectancy.data_loading import load_data, save_data
+from life_expectancy.data_cleaning import clean_data
+
+PROJECT_DIR = Path(__file__).parent
+FILE_PATH = f"{PROJECT_DIR}/data/eu_life_expectancy_raw.tsv"
```
Comment (MrCordeiro) [inline life_expectancy/main.py]:
```suggestion
FILE_PATH = PROJECT_DIR / "data" / "eu_life_expectancy_raw.tsv"
```

File: life_expectancy/data_loading.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"""
+Loading/saving script
+"""
+
+import pandas as pd
+
+def load_data(file_path: str, sep: str = ",") -> pd.DataFrame:
```
Comment (MrCordeiro) [inline life_expectancy/data_loading.py]:
```suggestion
def load_data(file_path: Path | str, sep: str = ",") -> pd.DataFrame:
```

Comment (MrCordeiro) [reply life_expectancy/data_loading.py]: Acho para Python < 3.10, usas:

from typings import Union

Union[str, Path]

File: life_expectancy/tests/conftest.py [hunk 4]
```diff
@@ -16,8 +20,14 @@ def run_before_and_after_tests() -> None:
     file_path = OUTPUT_DIR / "pt_life_expectancy.csv"
     file_path.unlink(missing_ok=True)
```
Comment (MrCordeiro) [inline life_expectancy/tests/conftest.py:22]: **Comment** Podemos deletar este setup se tudo estiver ok

File: life_expectancy/tests/test_cleaning.py
Comment (MrCordeiro) [inline life_expectancy/tests/test_cleaning.py:1]: Criar um module de testes para cada modulo do projeto

File: life_expectancy/tests/test_cleaning.py [hunk 5]
```diff
@@ -1,18 +1,45 @@
...
+
+def test_load_data(eu_life_expectancy_raw):
+    """Test for the loading data function"""
+
+    dataframe = load_data(FILE_PATH, "\t")
+    pd.testing.assert_frame_equal(
+        dataframe, eu_life_expectancy_raw
+    )
+
+@mock.patch("pandas.DataFrame.to_csv")
```
Comment (MrCordeiro) [inline life_expectancy/tests/test_cleaning.py]:
```suggestion
@mock.patch("life_expectancy.data_loading.pd.DataFrame.to_csv")
```

File: life_expectancy/tests/test_cleaning.py [hunk 6]
```diff
@@ -1,18 +1,45 @@
...
-from . import OUTPUT_DIR
+from life_expectancy.main import main
+from life_expectancy.data_loading import load_data, save_data
+from life_expectancy.data_cleaning import clean_data
 
+PROJECT_DIR = Path(__file__).parent.parent
+FILE_PATH = f"{PROJECT_DIR}/data/eu_life_expectancy_raw.tsv"
 
-def test_clean_data(pt_life_expectancy_expected):
+def test_main(pt_life_expectancy_expected):
```
Comment (MrCordeiro) [inline life_expectancy/tests/test_cleaning.py]: Faltam os mocks para ler (para consumir teu fixture) e gravar dataframe.

Comment (MrCordeiro) [reply life_expectancy/tests/test_cleaning.py]:
```python
@patch(life_expactancy.main.read_data)
@patch(save)
def test_xxxx (mock_save, mock_read, ...)

   # Para que o mock retorne a minha fixture
   mock_read.return_value = minha_fixture
   
   # Para que o mock realize alguma ação ao ser chamado
   mock_save.side_effect = print("I'm saving")
```

**Review by MrCordeiro — CHANGES_REQUESTED** [review]

File: life_expectancy/main.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+SAVE_PATH = "life_expectancy/data/pt_life_expectancy.csv"
```
Comment (filipacarreira) [inline life_expectancy/main.py]: this variable can't also use the PROJECT_DIR variable?
Like: `SAVE_PATH = PROJECT_DIR / data / pt_life_expectancy.csv`

File: life_expectancy/tests/test_cleaning.py [hunk 8]
```diff
@@ -1,18 +1,45 @@
...
 import pandas as pd
 
-from life_expectancy.cleaning import main
-from . import OUTPUT_DIR
+from life_expectancy.main import main
+from life_expectancy.data_loading import load_data, save_data
+from life_expectancy.data_cleaning import clean_data
 
+PROJECT_DIR = Path(__file__).parent.parent
+FILE_PATH = f"{PROJECT_DIR}/data/eu_life_expectancy_raw.tsv"
```
Comment (filipacarreira) [inline life_expectancy/tests/test_cleaning.py]: Is the same suggestion Fernando did before, changing it to
`FILE_PATH = PROJECT_DIR / "data" / "eu_life_expectancy_raw.tsv"`

**Review by filipacarreira — COMMENTED** [review]
Just commented minor changes you could implement, but really great job André! Congrats 🥳

File: life_expectancy/tests/test_loading.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+from life_expectancy.data_loading import load_data, save_data
+
+PROJECT_DIR = Path(__file__).parent.parent
+FILE_PATH = f"{PROJECT_DIR}/data/eu_life_expectancy_raw.tsv"
+
+def test_load_data(eu_life_expectancy_raw):
+    """Test for the loading data function"""
+
+    dataframe = load_data(FILE_PATH, "\t")
```
Comment (MrCordeiro) [inline life_expectancy/tests/test_loading.py:13]: Aqui deverias estar a consumir das fixtures, não do directorio `data`

**Review by MrCordeiro — CHANGES_REQUESTED** [review]

**Review by MrCordeiro — APPROVED** [review]

**Pull request merged**